File: mac_old/Sources/FileGuilei/Services/FileOrganizer.swift
import Foundation

@Observable
class FileOrganizer {
    enum State: Equatable {
        case idle
        case copying(progress: Double, currentFile: String)
        case waitingConfirmation
        case deleting(progress: Double, currentFile: String)
        case completed(fileCount: Int, totalSize: Int64)
        case failed(String)

        static func == (lhs: State, rhs: State) -> Bool {
            switch (lhs, rhs) {
            case (.idle, .idle): true
            case (.copying(let lp, _), .copying(let rp, _)): lp == rp
            case (.waitingConfirmation, .waitingConfirmation): true
            case (.deleting(let lp, _), .deleting(let rp, _)): lp == rp
            case (.completed, .completed): true
            case (.failed(let lm), .failed(let rm)): lm == rm
            default: false
            }
        }
    }

    enum Operation {
        case createdDirectory(URL)
        case copiedFile(source: URL, destination: URL)
    }

    var state: State = .idle
    private var operationLog: [Operation] = []
    private let fm = FileManager.default

    func copyFiles(scanResult: ScanResult) async throws {
        operationLog.removeAll()
        let selectedFiles = scanResult.categorizedFiles.flatMap { category, files in
            files.filter(\.isSelected).map { (category, $0) }
        }
        let total = selectedFiles.count
        guard total > 0 else { return }

        for (index, (category, file)) in selectedFiles.enumerated() {
            let progress = Double(index) / Double(total)
            await MainActor.run { state = .copying(progress: progress, currentFile: file.name) }

            let categoryDir = scanResult.sourceDirectory.appending(path: category.displayName)
            if !fm.fileExists(atPath: categoryDir.path) {
                try fm.createDirectory(at: categoryDir, withIntermediateDirectories: true)
                operationLog.append(.createdDirectory(categoryDir))
            }

            let destination = uniqueDestination(for: file.name, in: categoryDir)
            do {
                try fm.copyItem(at: file.url, to: destination)
                operationLog.append(.copiedFile(source: file.url, destination: destination))
            } catch {
                try await rollback()
                await MainActor.run { state = .failed(error.localizedDescription) }
                throw error
            }
        }
        await MainActor.run { state = .waitingConfirmation }
    }

    func deleteOriginals(scanResult: ScanResult) async throws {
        let selectedFiles = scanResult.categorizedFiles.values.flatMap { $0 }.filter(\.isSelected)
        let total = selectedFiles.count
        guard total > 0 else { return }
        for (index, file) in selectedFiles.enumerated() {
            let progress = Double(index) / Double(total)
            await MainActor.run { state = .deleting(progress: progress, currentFile: file.name) }
            try fm.removeItem(at: file.url)
        }
        let totalSize = selectedFiles.reduce(0) { $0 + $1.size }
        await MainActor.run { state = .completed(fileCount: total, totalSize: totalSize) }
    }

    func rollback() async throws {
        for operation in operationLog.reversed() {
            switch operation {
            case .copiedFile(_, let destination):
                try? fm.removeItem(at: destination)
            case .createdDirectory(let dir):
                let contents = try? fm.contentsOfDirectory(at: dir, includingPropertiesForKeys: nil)
                if contents?.isEmpty == true { try? fm.removeItem(at: dir) }
            }
        }
        operationLog.removeAll()
        await MainActor.run { state = .idle }
    }

    func reset() {
        operationLog.removeAll()
        state = .idle
    }

    private func uniqueDestination(for filename: String, in directory: URL) -> URL {
        let base = (filename as NSString).deletingPathExtension
        let ext = (filename as NSString).pathExtension
        var candidate = directory.appending(path: filename)
        var counter = 1
        while fm.fileExists(atPath: candidate.path) {
            let newName = ext.isEmpty ? "\(base) (\(counter))" : "\(base) (\(counter)).\(ext)"
            candidate = directory.appending(path: newName)
            counter += 1
        }
        return candidate
    }
}

File: allplatform/src/main/services/FileOrganizer.ts
import fs from 'fs'
import path from 'path'
import { FileCategory, ScanResult, OrganizerState, Operation } from '../../shared/types'
import { getCategoryDisplayName } from '../models/FileCategory'

export class FileOrganizer {
  private operationLog: Operation[] = []
  private state: OrganizerState = { phase: 'idle' }
  private sendProgress: ((state: OrganizerState) => void) | null = null
  private copyMap: Map<string, string> = new Map()

  setSendProgress(fn: (state: OrganizerState) => void): void {
    this.sendProgress = fn
  }

  getState(): OrganizerState {
    return this.state
  }

  getCopyMap(): Map<string, string> {
    return this.copyMap
  }

  logCreatedFile(filePath: string): void {
    this.operationLog.push({ type: 'createdFile', path: filePath })
  }

  logCreatedDirectory(dirPath: string): void {
    this.operationLog.push({ type: 'createdDirectory', path: dirPath })
  }

  private setState(newState: OrganizerState): void {
    this.state = newState
    if (this.sendProgress) this.sendProgress(newState)
  }

  async copyFiles(scanResult: ScanResult): Promise<void> {
    this.operationLog = []
    this.copyMap = new Map()
    const selectedFiles: { category: FileCategory; file: { id: string; path: string; name: string } }[] = []

    for (const [category, files] of Object.entries(scanResult.categorizedFiles)) {
      if (!files) continue
      for (const file of files) {
        if (!file.isSelected) continue
        if (file.fromDeepScan) continue
        if ((category as FileCategory) === FileCategory.Folder) continue
        selectedFiles.push({ category: category as FileCategory, file: { id: file.id, path: file.path, name: file.name } })
      }
    }

    const total = selectedFiles.length
    if (total === 0) {
      this.setState({ phase: 'waitingConfirmation' })
      return
    }

    for (let i = 0; i < total; i++) {
      const { category, file } = selectedFiles[i]
      const progress = i / total
      this.setState({ phase: 'copying', progress, currentFile: file.name })

      const categoryDir = path.join(scanResult.sourceDirectory, getCategoryDisplayName(category))
      if (!fs.existsSync(categoryDir)) {
        fs.mkdirSync(categoryDir, { recursive: true })
        this.operationLog.push({ type: 'createdDirectory', path: categoryDir })
      }

      const destination = this.uniqueDestination(file.name, categoryDir)
      try {
        fs.copyFileSync(file.path, destination)
        this.operationLog.push({ type: 'copiedFile', source: file.path, destination })
        this.copyMap.set(file.id, destination)
      } catch (err) {
        await this.rollback()
        this.setState({ phase: 'failed', error: (err as Error).message })
        throw err
      }
    }

    this.setState({ phase: 'waitingConfirmation' })
  }

  async deleteOriginals(scanResult: ScanResult): Promise<void> {
    const selectedFiles = Object.values(scanResult.categorizedFiles)
      .flat()
      .filter(f => f && f.isSelected && !f.fromDeepScan && f.category !== FileCategory.Folder)

    const total = selectedFiles.length
    if (total === 0) return

    let totalSize = 0
    for (let i = 0; i < total; i++) {
      const file = selectedFiles[i]!
      this.setState({ phase: 'deleting', progress: i / total, currentFile: file.name })
      fs.unlinkSync(file.path)
      totalSize += file.size
    }

    this.setState({ phase: 'completed', fileCount: total, totalSize })
  }

  async rollback(): Promise<void> {
    for (const op of [...this.operationLog].reverse()) {
      if (op.type === 'copiedFile') {
        try { fs.unlinkSync(op.destination) } catch { /* already removed */ }
      } else if (op.type === 'createdFile') {
        try { fs.unlinkSync(op.path) } catch { /* already removed */ }
      } else if (op.type === 'createdDirectory') {
        try {
          const contents = fs.readdirSync(op.path)
          if (contents.length === 0) fs.rmdirSync(op.path)
        } catch { /* already removed */ }
      }
    }
    this.operationLog = []
    this.copyMap = new Map()
    this.setState({ phase: 'idle' })
  }

  reset(): void {
    this.operationLog = []
    this.copyMap = new Map()
    this.state = { phase: 'idle' }
  }

  private uniqueDestination(filename: string, directory: string): string {
    const ext = path.extname(filename)
    const base = path.basename(filename, ext)
    let candidate = path.join(directory, filename)
    let counter = 1
    while (fs.existsSync(candidate)) {
      const newName = ext ? `${base} (${counter})${ext}` : `${base} (${counter})`
      candidate = path.join(directory, newName)
      counter++
    }
    return candidate
  }
}
